Add MongoDB guide, MySQL Cloud SQL guide and other 7.0 docs updates (#7350)
gravitational/teleport: docs/pages/database-access/guides/mongodb-self-hosted.mdx

---
title: Database Access with Self-Hosted MongoDB
description: How to configure Teleport Database Access with self-hosted MongoDB
---

# Self-Hosted MongoDB

<Admonition type="note" title="Supported versions">
  Teleport Database Access supports MongoDB 3.6 and higher. Older versions have
  not been tested and are not guaranteed to work. MongoDB 3.6 was released in
  November 2017 and reached EOL in April 2021 so if you're still using an older
  version, consider upgrading.
</Admonition>

## Create Certificate/Key Pair

(!docs/pages/includes/database-access/tctl-auth-sign.mdx!)

Create the secrets:

<Tabs>
  <TabItem label="Standalone">
  When connecting to standalone MongoDB, sign certificate for the hostname over
  which Teleport will be connecting to it.

  For example, if your MongoDB server is accessible at `mongo.example.com`
  hostname, run:

  ```bash

>   $ tctl auth sign --format=mongodb --host=mongo.example.com --out=mongo --ttl=8760h

Reviewer: Double-checking: we're using `8760h` here but `8192h` elsewhere in this PR. Is there one value we should use? Are these values specific to one of the DB's but not the others?

---
title: Database Access with Self-Hosted MongoDB
description: How to configure Teleport Database Access with self-hosted MongoDB
---

# Self-Hosted MongoDB

<Admonition type="note" title="Supported versions">
  Teleport Database Access supports MongoDB 3.6 and higher. Older versions have
  not been tested and are not guaranteed to work. MongoDB 3.6 was released in
  November 2017 and reached EOL in April 2021 so if you're still using an older
  version, consider upgrading.
</Admonition>

## Create Certificate/Key Pair

(!docs/pages/includes/database-access/tctl-auth-sign.mdx!)

Create the secrets:

<Tabs>
  <TabItem label="Standalone">

>   When connecting to standalone MongoDB, sign certificate for the hostname over

Reviewer: May want to double-check this component. Looks like a tag is unclosed.

Author: It is closed, renders fine also.

command:

```js
db.getSiblingDB("$external").runCommand(
  {
    createUser: "CN=alice",
    roles: [
      { role: "readWriteAnyDatabase", db: "admin" }
    ]
  }
)
```

Update the [roles](https://docs.mongodb.com/manual/tutorial/manage-users-and-roles/)
accordingly to grant the user appropriate database permissions.

See [Use X.509 Certificates to Authenticate Clients](https://docs.mongodb.com/manual/tutorial/configure-x509-client-authentication/)
in MongoDB documentation for more details.

## Configure MongoDB TLS

Add the following section to your `mongod.conf` configuration file to enable
mutual TLS. Use the files generated above by the `tctl auth sign` command:

<Tabs>
  <TabItem label="MongoDB 3.6 - 4.2">
  ```yaml
  net:
    ssl:
      mode: requireSSL
      PEMKeyFile: /etc/certs/mongo.pem
      CAFile: /etc/certs/mongo.cas
  ```
  </TabItem>
  <TabItem label="MongoDB 4.2+">
  ```yaml
  net:
    tls:
      mode: requireTLS
      certificateKeyFile: /etc/certs/mongo.pem
      CAFile: /etc/certs/mongo.cas
  ```
  </TabItem>
</Tabs>

When configuring a replica set, make sure to do it for each member and use
secrets generated for the particular server.

> <Admonition type="note" title="">

Reviewer: Stylistic comment: I think `title` is optional, you can either leave it out or add whatever you want. We often will just add the `type`. 

Thanks for using these!

  Once mutual TLS has been enabled, you will no longer be able to connect to
  the cluster without providing a valid client certificate. You can use
  `net.tls.allowConnectionsWithoutCertificates` setting to allow connections
  from clients that do not present a certificate.
</Admonition>

See [Configure TLS/SSL](https://docs.mongodb.com/manual/tutorial/configure-ssl/)
in MongoDB documentation for more details.

## Setup Teleport Auth and Proxy Services

Teleport Database Access for MongoDB is available starting from `7.0` release.

> (!docs/pages/includes/database-access/start-auth-proxy.mdx!)

Reviewer: Thanks for refactoring some of these into includes!

  like to connect to a secondary instead, Teleport will respect `readPreference`
  connection string setting:

  ```yaml
  uri: "mongodb://mongo1.example.com:27017,mongo2.example.com:27017/?replicaSet=rs0&readPreference=secondary"
  ```

  Note that if leader election is triggered during the session and the member
  becomes primary, you won't be auto-reconnected.
  </TabItem>
</Tabs>

## Connect

Once the database service has joined the cluster, login to see the available
databases:

```sh
$ tsh login --proxy=teleport.example.com:3080 --user=testuser
$ tsh db ls
Name          Description     Labels
------------- --------------- --------
example-mongo Example MongoDB env=dev
```

Note that you will only be able to see databases your role has access to. See
[RBAC](../rbac.mdx) section for more details.

To connect to a particular database server, first retrieve credentials from
Teleport using `tsh db login` command:

```sh
$ tsh db login example-mongo
```

<Admonition type="tip" title="Tip">
  You can be logged into multiple databases simultaneously.
</Admonition>

You can optionally specify the database name and the user to use by default
when connecting to the database instance:

> ```sh

Reviewer: I think `bash` and `sh` render the same but not `bsh`. As you know, the values we populate in the code blocks determines the coloring when they render. I've gone through and tried to use `bash` universally just to be sure.

It'll also help us to quickly GREP on all code snippets in the future.

allow rule in their role:

```yaml
allow:
  impersonate:
    users: ["Db"]
    roles: ["Db"]
```

Create the secrets:

<Tabs>
  <TabItem label="Standalone">
  When connecting to standalone MongoDB, sign certificate for the hostname over
  which Teleport will be connecting to it.

  For example, if your MongoDB server is accessible at `mongo.example.com`
  hostname, run:

  ```bash
  $ tctl auth sign --format=mongodb --host=mongo.example.com --out=mongo --ttl=8760h
  ```

  The command will create 2 files: `mongo.cas` with Teleport's certificate
  authority and `mongo.crt` with generated certificate and key pair. You will
  need these files to enable mutual TLS on your MongoDB server.
  </TabItem>
  <TabItem label="Replica set">
  When connecting to a MongoDB replica set, sign certificates for each member
  using the hostnames they're accessible at.

  For example, if the first member is accessible at `mongo1.example.com` and
  the second at `mongo2.example.com`, run:

  ```bash
  $ tctl auth sign --format=mongodb --host=mongo1.example.com --out=mongo1 --ttl=8760h
  $ tctl auth sign --format=mongodb --host=mongo2.example.com --out=mongo2 --ttl=8760h
  ```

>   Each command will create 2 files: `mongo{1,2}.cas` with Teleport's certificate

Reviewer 2: ```suggestion
  Each command will create 2 files: `mongo1.cas` and `mongo2.cas` with Teleport's certificate
```

Might be better to be explicit about the filenames here - some people may not be familiar with the `bash` `{1,2}` syntax format.

</Tabs>

When configuring a replica set, make sure to do it for each member and use
secrets generated for the particular server.

<Admonition type="note" title="">
  Once mutual TLS has been enabled, you will no longer be able to connect to
  the cluster without providing a valid client certificate. You can use
  `net.tls.allowConnectionsWithoutCertificates` setting to allow connections
  from clients that do not present a certificate.
</Admonition>

See [Configure TLS/SSL](https://docs.mongodb.com/manual/tutorial/configure-ssl/)
in MongoDB documentation for more details.

## Configure Teleport

Teleport Database Access for MongoDB is available starting from `7.0` release.

Download the appropriate version of Teleport for your platform from
our [downloads page](https://goteleport.com/teleport/download).

Follow the installation [instructions](../../installation.mdx).

### Start Auth/Proxy Service

Create a configuration file for a Teleport service that will be running
auth and proxy servers:

```yaml
teleport:
  data_dir: /var/lib/teleport
  nodename: test
auth_service:
  enabled: "yes"
proxy_service:
  enabled: "yes"
  # Set public address proxy will be reachable at.
  #

>   # MongoDB clients are connecting to regular web proxy endpoint.

Reviewer 2: ```suggestion
  # MongoDB clients connect to Teleport's regular web proxy endpoint.
```

  public_addr: teleport.example.com:3080
ssh_service:
  enabled: "no"
```

Start the service:

```sh
$ teleport start --config=/path/to/teleport.yaml
```

Generate a short-lived join token for the database service and save it for
example in `/tmp/token`:

```sh
$ tctl tokens add --type=db
```

### Create Role and User

> Create the role that will allow a user to connect to any database using any

Reviewer 2: ```suggestion
Create a role that will allow a user to connect to any database using any
```

database user:

```sh
$ tctl --config=/path/to/teleport-db-role.yaml create <<EOF
kind: role
version: v3
metadata:
  name: db
spec:
  allow:
    db_labels:
      '*': '*'
    db_names:
    - '*'
    db_users:
    - '*'
EOF
```

> Create the user assigned the `db` role we've just created:

Reviewer 2: ```suggestion
Create a user assigned the `db` role we've just created:
```
